src/asm.rs

```diff
@@ -0,0 +1,268 @@
+/// safety requirements.
+#[inline]
+#[cfg_attr(feature = "on_gba", instruction_set(arm::a32))]
+pub unsafe fn a32_load_i16_reg_offset(ptr: *mut i16, bytes: usize) -> i16 {
+  on_gba_or_unimplemented! {
+    let output: i16;
+    unsafe {
+      core::arch::asm! {
+        "ldrsh {output}, [{ptr}, {bytes}]",
+        output = lateout(reg) output,
+        ptr = in(reg) ptr,
+        bytes = in(reg) bytes,
+      }
+    }
+    output
+  }
+}
+
+// Proc-macros can't see the target being built for, so we use this declarative
+// macro to determine if we're on a thumb target (and need to force our asm into
+// a32 mode) or if we're not on thumb (and our asm can pass through untouched).
+#[cfg(target_feature = "thumb-mode")]
+macro_rules! force_a32 {
+  ($($asm_line:expr),+ $(,)?) => {
+    t32_with_a32_scope! {
+      $( concat!($asm_line, "\n") ),+ ,
+    }
+  }
+}
+#[cfg(not(target_feature = "thumb-mode"))]
+macro_rules! force_a32 {
+  ($($asm_line:expr),+ $(,)?) => {
+    concat!(
+      $( concat!($asm_line, "\n") ),+ ,
+    )
+  }
+}
+
+#[cfg(feature = "on_gba")]
+core::arch::global_asm! {
+  put_fn_in_section!(".text._start"),
+  ".global _start",
+  force_a32!{
+    "_start:",
+    // space for the rom header to be placed after compilation.
+    "b 1f",
+    ".space 0xE0",
+    "1:",
+
+    "mov r12, #0x04000000",
+    "add r3, r12, #0xD4",
+
+    // Now
+    // * r12: mmio base
+    // * r3: dma3 base
+
+    // Configure WAITCNT to the GBATEK suggested default
+    "add r0, r12, #0x204",
+    "ldr r1, =0x4317",
+    "strh r1, [r0]",
+
+    // activate mGBA logging, when available
+    "ldr r0, =0x04FFF780",
+    "ldr r1, =0xC0DE",
+    "strh r1, [r0]",
+
+    /* iwram copy */
+    "ldr r0, =_iwram_word_copy_count",
+    when!(("r0" != "#0")[1] {
+      "ldr r1, =_iwram_position_in_rom",
+      "str r1, [r3]",           // src
+      "ldr r1, =_iwram_start",
+      "str r1, [r3, #4]",       // dest
+      "strh r0, [r3, #8]",      // transfers
+      "mov r1, #(1<<10|1<<15)", // 32-bit transfers, enable
+      "strh r1, [r3, #10]",
+    }),
+
+    /* ewram copy */
+    "ldr r4, =_ewram_word_copy_count",
+    when!(("r4" != "#0")[1] {
+      "ldr r1, =_ewram_position_in_rom",
+      "str r1, [r3]",
+      "ldr r1, =_ewram_start",
+      "str r1, [r3, #4]",
+      "strh r0, [r3, #8]",
+      "mov r1, #(1<<10|1<<15)",
+      "strh r1, [r3, #10]",
+    }),
+
+    /* bss zero */
+    "ldr r4, =_bss_word_clear_count",
+    when!(("r4" != "#0")[1] {
+      "ldr r0, =_bss_start",
+      "mov r2, #0",
+      "2:",
+      "str r2, [r0], #4",
+      "subs r4, r4, #1",
+      "bne 2b",
+    }),
+
+    // Tell the BIOS about our irq handler
+    "ldr r0, =_asm_runtime_irq_handler",
+    "str r0, [r12, #-4]",
+
+    // Note(Lokathor): we do a `bx` instead of a `b` because it saves 4 *entire*
+    // bytes (!), since `main` will usually be a t32 function and thus usually
+    // requires a linker shim to call.
+    "ldr r0, =main",
+    "bx r0",
```

Comment:
`bx r0` will branch-exchange to the fn pointer in r0. this is the instruction that allows a code state switch as part of the call.

`b main` (branch) is an alternative way to do this call. in this case, it would directly jump to main and we would have a problem if the code state didn't match.

`bl main` (branch-link) is another way as well. With `bl`, the linker would check the code state of the caller and callee, and make the caller call to the switching shim if the caller and callee don't match. This is a few cycles slower and probably wastes 4 bytes of rom size. It doesn't really matter to save the cycles or code size in this case since it's once per boot, but we might as well.

Reply:
> I notice also that there's no explicit stack setup going on here. Given that the stack doesn't seem to have reserved space in the link script, does this means that too many things in IWRAM will result in corruption due to the stack (I believe it's set up at the end of IWRAM by default, right?).

Correct, the GBA doesn't have any sort of stack protection mechanism. I'm gonna resolve this sub-thread and make this a todo in the top level post.